lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,881 @@
+			&responder_static_public_key,
+			chaining_key.into_inner(),
+			hash,
+			&mut act_one
+		);
+
+		Ok((
+			Some(Act::One(act_one)),
+			HandshakeState::InitiatorAwaitingActTwo(InitiatorAwaitingActTwoState {
+				initiator_static_private_key,
+				initiator_static_public_key,
+				initiator_ephemeral_private_key,
+				responder_static_public_key,
+				chaining_key,
+				hash,
+				act_two_builder: ActBuilder::new(Act::Two(EMPTY_ACT_TWO))
+			})
+		))
+	}
+}
+
+impl ResponderAwaitingActOneState {
+	pub(crate) fn new(responder_static_private_key: SecretKey, responder_ephemeral_private_key: SecretKey) -> Self {
+		let responder_static_public_key = private_key_to_public_key(&responder_static_private_key);
+		let (hash, chaining_key) = handshake_state_initialization(&responder_static_public_key);
+		let responder_ephemeral_public_key = private_key_to_public_key(&responder_ephemeral_private_key);
+
+		ResponderAwaitingActOneState {
+			responder_static_private_key,
+			responder_ephemeral_private_key,
+			responder_ephemeral_public_key,
+			chaining_key,
+			hash,
+			act_one_builder: ActBuilder::new(Act::One(EMPTY_ACT_ONE))
+		}
+	}
+}
+
+impl IHandshakeState for ResponderAwaitingActOneState {
+	// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (receiver)
+	// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-two (sender)
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		let mut act_one_builder = self.act_one_builder;
+		let remaining = act_one_builder.fill(input);
+
+		// Any payload larger than ACT_ONE_TWO_LENGTH indicates a bad peer since initiator data
+		// is required to generate act3 (so it can't come before we transition)
+		if remaining.len() != 0 {
+			return Err("Act One too large".to_string());
+		}
+
+		// In the event of a partial fill, stay in the same state and wait for more data
+		if !act_one_builder.is_finished() {
+			assert_eq!(remaining.len(), 0);
+			return Ok((
+				None,
+				HandshakeState::ResponderAwaitingActOne(Self {
+					responder_static_private_key: self.responder_static_private_key,
+					responder_ephemeral_private_key: self.responder_ephemeral_private_key,
+					responder_ephemeral_public_key: self.responder_ephemeral_public_key,
+					chaining_key: self.chaining_key,
+					hash: self.hash,
+					act_one_builder
+				})
+			));
+		}
+
+		let hash = self.hash;
+		let responder_static_private_key = self.responder_static_private_key;
+		let chaining_key = self.chaining_key;
+		let responder_ephemeral_private_key = self.responder_ephemeral_private_key;
+		let responder_ephemeral_public_key = self.responder_ephemeral_public_key;
+		let act_one = Act::from(act_one_builder);
+
+		let (initiator_ephemeral_public_key, hash, chaining_key, _) = process_act_message(
+			&act_one,
+			&responder_static_private_key,
+			chaining_key.into_inner(),
+			hash,
+		)?;
+
+		let mut act_two = EMPTY_ACT_TWO;
```

Comment:
For ciphertext version, can we explicitly set it, either here or in `calculate_act_message`, to dedup zero'ing from version setting even if it doesn't matter for now ?

Reply:
Resolved in `review: Explicitly set version in creation`